### DISCUSSION TITLE
Cross platform determinism issue (Windows - Linux)

### DISCUSSION THREAD

**Discussion by AndreaCatania** [jrouwe/JoltPhysics#1413] [General]
Hello,

I am encountering a problem in maintaining Jolt determinism between two different platforms, Windows and Linux, where the client runs the game on Windows and the server instead on Linux.

After several tests and having disabled all the game code to ensure that the determinism issue is not caused by the game, updating Jolt to the latest MASTER version, and making sure not to use SIN/COS/ATAN/sort, I have observed that Jolt does not behave deterministically. A small note: when both the client and the server use Windows as the platform, everything works perfectly.

To provide some context, the game is implemented in Unreal Engine, and Jolt is compiled as part of a CMake library that contains both the gameplay code and the state synchronization code alongside Jolt. This library is statically linked to various Unreal Engine modules (DLLs), which read states and positions to update the positions of the various objects on the screen. This setup ensures that all the gameplay code remains separate from Unreal Engine, allowing complete control over the execution flow as well as the compile flags, enabling the setting of various flags like /fp:precise, etc.

The synchronization problem appears somewhat everywhere, but in my test setup, I am focusing on the VirtualCharacter (Kinematic). What I do is process the character with hardcoded velocity and direction to ensure there are no external interferences. Nevertheless, consistently, the buffer produced by the client differs from the buffer produced by the server.

Have you had similar experiences?
What can I do to identify the problem?

<details>
<summary>Here is the CMake code I use to compile Jolt:</summary>

```cmake
cmake_minimum_required(VERSION 3.13)
project(Flash CXX)

set(CMAKE_C_STANDARD 11)
set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_STANDARD_REQUIRED True)

# Set repository root
set(FLASH_REPO_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/../Source)

#                                                                              #
# Options ------------------------------------------------------------ Options #
#                                                                              #

# When turning this option on, the library will be compiled with debug symbols
option(GENERATE_DEBUG_SYMBOLS "Generate debug symbols" OFF)

# When turning this option on, the library will be compiled with extra checks and assertions.
option(ENABLE_DEBUG "Enable debug checks" OFF)

option(ENABLE_EDITOR "Enable editor code, defines the macro FLASH_EDITOR." OFF)

# When turning this option on, you can override the release and debug CPP build flags.
option(OVERRIDE_CXX_FLAGS "Enable this to override the flags used for release and debug. `CMAKE_CXX_FLAGS_DEBUG` `CMAKE_CXX_FLAGS_RELEASE`" OFF)

# Ability to toggle between the static and DLL versions of the MSVC runtime library
# Windows Store only supports the DLL version
option(USE_STATIC_MSVC_RUNTIME_LIBRARY "Use the static MSVC runtime library" ON)

# Ability to toggle between the static and DLL versions of the MSVC runtime library
# Windows Store only supports the DLL version
option(BUILD_MSVC_MD "In MSVC build the library using MD." ON)

#                                                                              #
# Section: -------------------------------------------- Third-Party Dependencies
#                                                                              #
# ------------------------------------------------------------------------- Jolt
option(JOLT_USE_ASSERTS "Use Jolt assetion." OFF)

function(add_jolt)
    set(USE_ASSERTS ${JOLT_USE_ASSERTS} CACHE BOOL "Controls the jolt assertion." FORCE)
    
    set(CROSS_PLATFORM_DETERMINISTIC ON CACHE BOOL "Jolt - Always enable crossplatform determinism - Essentially uses precise float." FORCE)
    set(DOUBLE_PRECISION ON CACHE BOOL "Jolt - double precision" FORCE)
    set(OBJECT_LAYER_BITS 16 CACHE BOOL "Jolt - Layer bits" FORCE)
    set(INTERPROCEDURAL_OPTIMIZATION OFF CACHE BOOL "Jolt - Link time opt" FORCE)
    set(CMAKE_EXPORT_COMPILE_COMMANDS ON CACHE BOOL "Jolt - Export commands." FORCE)
    set(TARGET_SAMPLES OFF CACHE BOOL "Jolt - Generate samples." FORCE)
    set(TARGET_HELLO_WORLD OFF CACHE BOOL "Jolt - Generate hello world." FORCE)
    set(TARGET_VIEWER OFF CACHE BOOL "Jolt - target debug viewer." FORCE)
    set(TARGET_UNIT_TESTS OFF CACHE BOOL "Jolt - Unit tests." FORCE)
    set(TARGET_PERFORMANCE_TEST OFF CACHE BOOL "Jolt - Perf tests." FORCE)

    # Disable warnings otherwise this problem occurs: https://developercommunity.visualstudio.com/t/Including-limitsh-triggers-error-__STDC/10514743?sort=newest 
    set(ENABLE_ALL_WARNINGS OFF CACHE BOOL "Disable all warning - NOTE this is set to OFF to avoid trigger a MSVC bug: https://developercommunity.visualstudio.com/t/Including-limitsh-triggers-error-__STDC/10514743?sort=newest" FORCE)
    
    set(JOLT_SIMD_ENABLE_AVX OFF)
    set(JOLT_SIMD_ENABLE_OTHERS ON) # TODO turn this to on, should not cause any determinism problem.
    set(USE_AVX512 ${JOLT_SIMD_ENABLE_AVX} CACHE BOOL "Jolt - Simd AVX512" FORCE)
    set(USE_AVX2 ${JOLT_SIMD_ENABLE_AVX} CACHE BOOL "Jolt - Simd AVX2" FORCE)
    set(USE_AVX ${JOLT_SIMD_ENABLE_AVX} CACHE BOOL "Jolt - Simd AVX" FORCE)
    set(USE_FMADD ${JOLT_SIMD_ENABLE_OTHERS} CACHE BOOL "Jolt - Simd FMADD" FORCE)
    set(USE_SSE4_2 ${JOLT_SIMD_ENABLE_OTHERS} CACHE BOOL "Jolt - Simd SSE4_2" FORCE)
    set(USE_SSE4_1 ${JOLT_SIMD_ENABLE_OTHERS} CACHE BOOL "Jolt - Simd SSE4_1" FORCE)
    set(USE_LZCNT ${JOLT_SIMD_ENABLE_OTHERS} CACHE BOOL "Jolt - Simd LZCNT" FORCE)
    set(USE_TZCNT ${JOLT_SIMD_ENABLE_OTHERS} CACHE BOOL "Jolt - Simd TZCNT" FORCE)
    set(USE_F16C ${JOLT_SIMD_ENABLE_OTHERS} CACHE BOOL "Jolt - Simd f16c" FORCE)
    
    # Adds the JoltPhysics library.
    add_subdirectory(
        ${FLASH_REPO_ROOT}/JoltPhysics/Build
        ${CMAKE_CURRENT_BINARY_DIR}/JoltPhysics
    )
endfunction()

add_jolt()

# --------------------------------------------------------- Network Synchronizer
option(NS_ENABLE_DEBUGGER_UI "Enables the debugger UI on the NetworkSynchronizer library." OFF)
function(add_network_synchronizer)
    set(ENABLE_DEBUGGER_UI ${NS_ENABLE_DEBUGGER_UI} CACHE BOOL "NetworkSynchronizer - Controls the NetworkSynchronizer UI debugger usage." FORCE)
    
    # Adds the NetworkSynchronizer library.
    add_subdirectory(
        ${FLASH_REPO_ROOT}/NetworkSynchronizer/cmake
        ${CMAKE_CURRENT_BINARY_DIR}/NetworkSynchronizer
    )
endfunction()

add_network_synchronizer()

# --------------------------------------------------------------------- Mimalloc

function(add_mimalloc)
    set(MI_BUILD_STATIC ON CACHE BOOL "Mimalloc - Build mimalloc as a static library." FORCE)
    set(MI_BUILD_SHARED OFF CACHE BOOL "Mimalloc - Disable mimalloc build shared." FORCE)
    set(MI_BUILD_OBJECT OFF CACHE BOOL "Mimalloc - Disable mimalloc build object." FORCE)
    set(MI_BUILD_TESTS OFF CACHE BOOL "Mimalloc - Disable mimalloc build tests." FORCE)
    set(MI_OVERRIDE OFF CACHE BOOL "Mimalloc - Disable mimalloc build override." FORCE)
    
    # Adds the Mimalloc library.
    add_subdirectory(
        ${FLASH_REPO_ROOT}/Mimalloc
        ${CMAKE_CURRENT_BINARY_DIR}/Mimalloc
    )
endfunction()

add_mimalloc()

#                                                                              #
# Section: -------------------------------------------------------- LibraryFlash
#                                                                              #

if (MSVC)
    # Floating point is always precise to ensure determinism across platforms.
	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /fp:precise")
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /fp:precise")
    
    # Disable RTTI
	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /GR-")
			
	# Optionally generate debug symbols
	if (GENERATE_DEBUG_SYMBOLS)
		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zi")
	endif()
	
	# Set runtime library
	if (USE_STATIC_MSVC_RUNTIME_LIBRARY)
		set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
	endif()
	
	
	# Set linker flags
	if (GENERATE_DEBUG_SYMBOLS)
    	set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /DEBUG")
    endif()
    
    if (BUILD_MSVC_MD)
	    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /MD")
	    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /MD")
	endif()
	
    # Set compiler flags for various configurations
    if (NOT OVERRIDE_CXX_FLAGS)
    	set(CMAKE_CXX_FLAGS_DEBUG "/GS /Od /Ob0 /RTC1")
        if (BUILD_MSVC_MD)
    	    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /MDd")
	        set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} /MDd")
        endif()
    	set(CMAKE_CXX_FLAGS_RELEASE "/GS- /Gy /O2 /Oi /Ot")
    endif()
    set(CMAKE_CXX_FLAGS_DISTRIBUTION "${CMAKE_CXX_FLAGS_RELEASE}")
else()
    # Floating point is always precise to ensure determinism across platforms.
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-fast-math -frounding-math -fno-trapping-math")
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fno-fast-math -frounding-math -fno-trapping-math")
    
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -ffp-model=precise -msse2 -mfpmath=sse")
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -ffp-model=precise -msse2 -mfpmath=sse")
    
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -ffp-contract=off")
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -ffp-contract=off")
    
    # Disable RTTI
	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-rtti")

	if (GENERATE_DEBUG_SYMBOLS)
		set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /DEBUG")
	endif()
	
	if (NOT OVERRIDE_CXX_FLAGS)
		set(CMAKE_CXX_FLAGS_DEBUG "-g -O0")
		set(CMAKE_CXX_FLAGS_RELEASE "-O2")
	endif()
	set(CMAKE_CXX_FLAGS_DISTRIBUTION "${CMAKE_CXX_FLAGS_RELEASE}")
endif()

# Flash Root
set(FLASH_ROOT ${FLASH_REPO_ROOT}/Flash)

# Find all .cpp and .h files in the /Flash directory
file(GLOB_RECURSE FLASH_SRC_FILES
    "${FLASH_ROOT}/**.cpp"
    "${FLASH_ROOT}/**.h"
)

# Group source files
source_group(TREE ${FLASH_ROOT} FILES ${FLASH_SRC_FILES})

# Determinism Root
set(DETERMINISM_ROOT ${FLASH_REPO_ROOT}/Determinism)

# Find all .cpp and .h files in the /Determinism directory
file(GLOB_RECURSE DETERMINISM_SOURCE_FILES
    "${DETERMINISM_ROOT}/**.cpp"
    "${DETERMINISM_ROOT}/**.h"
)

# Group source files
source_group(TREE ${DETERMINISM_ROOT} FILES ${DETERMINISM_SOURCE_FILES})

# Combine all source files
list(APPEND FLASH_SRC_FILES ${DETERMINISM_SOURCE_FILES})

# Create Flash lib
add_library(Flash ${FLASH_SRC_FILES})

target_include_directories(Flash PUBLIC ${FLASH_REPO_ROOT})

# Setting enable asserts flag
if (ENABLE_DEBUG)
	target_compile_definitions(Flash PUBLIC FLASH_DEBUG)
endif()

# Setting enable editor flag
if (ENABLE_EDITOR)
	target_compile_definitions(Flash PUBLIC FLASH_EDITOR)
endif()

#
# Section: ------------------------------------------------------ Link libraries
#

# Link Flash against Jolt
target_link_libraries(Flash PUBLIC Jolt)

# Link Flash against NetworkSynchronizer
target_link_libraries(Flash PUBLIC NetworkSynchronizer)

# Link Flash against Mimalloc
target_link_libraries(Flash PUBLIC mimalloc-static)
```

</details>

<details>
<summary>VirtualCharacter processing CPP</summary>

```cpp

// DeltaTime is fixed to 1.f/60.f

void JoltBodyKinematic::MoveCharacter(float DeltaTime, const JPH::Vec3& VelocityInM, bool StickToFloor, const JPH::Quat& Orientation)
{
	if (DeltaTime <= FLT_EPSILON)
	{
		return;
	}

	if make_unlikely(!CharacterVirtual || !GetBody())
	{
		// Nothing to do.
		return;
	}


	// TODO Remove
	CharacterVirtual->SetRotation(JPH::Quat::sIdentity());
	CharacterVirtual->SetLinearVelocity(JPH::Vec3(0.f, -2.f, 0.f));

	CharacterHorizontalVelocity = CharacterVirtual->GetLinearVelocity();
	CharacterHorizontalVelocity.SetY(0.0);

	CVUpdateSettings.mWalkStairsStepUp = JPH::Vec3(0., 1.0f, 0.0);
	CVUpdateSettings.mStickToFloorStepDown = JPH::Vec3::sZero();

	const JPH::DefaultBroadPhaseLayerFilter BroadPhaseLayerFilter(
		get_scene().get_jolt_physics().GetObjectVsBroadPhaseLayerFilter(),
		layer);

	const JObjectLayerFilter LayerFilter(
		get_scene().get_jolt_physics().GetCollisionResponseTable(),
		layer);

	// Can collide with anything.
	const JPH::ShapeFilter ShapeFilter;

	CharacterVirtual->ExtendedUpdate(
		DeltaTime,
		get_scene().get_jolt_main_world().World_GetGravity(),
		CVUpdateSettings,
		BroadPhaseLayerFilter,
		LayerFilter,
		BodiesFilter,
		ShapeFilter,
		*get_scene().get_jolt_main_world().World_GetTempAllocator());

	// Wakeup the body to make sure it's processed.
	get_scene().get_jolt_main_world().Body_Activate(*this);

	// Move the Character to the location specified by the CharacterVirtual.
	const JPH::RVec3 Position = CharacterVirtual->GetPosition();

	GetBody()->MoveKinematic(
		Position,
		JPH::Quat::sIdentity(), // TODO remove
		//Orientation, // TODO use this!
		DeltaTime);

	const JPH::Vec3 TheActualVelocity = GetBody()->GetLinearVelocity();

	JPH::Vec3 V = VelocityInM;
	if (CharacterVirtual->IsSupported())
	{
		// This is disabled because would cause the character to gain speed when moving upward.
		//V.SetY(FMath::Max(VelocityInM.GetY(), Vel.GetY()));
	}
	else
	{
		// When not supported we expect being able to move up freely
		// unless it hits the roof. In this case cut the upward speed to avoid going up forever.
		V.SetY(std::min(VelocityInM.GetY(), TheActualVelocity.GetY()));
	}
	CharacterVirtual->SetLinearVelocity(V);
}
```

</details>

**Comment by jrouwe** [comment]
Hello,

These things are very hard to debug. What I do is to enable `JPH_ENABLE_DETERMINISM_LOG` which will start spitting out tons of TTY. I then sprinkle the code with things like:

```
JPH_DET_LOG("CharacterVirtual::MoveShape: pos: " << ioPosition << " vel: " << inVelocity << " dt: " << inDeltaTime);
```

This will print out all float values in hexadecimal format so that it is easy to spot 1 bit differences. I then use a text diff application to compare the text streams between platforms. When I see a difference, I add more `JPH_DET_LOG` statements until I see what's causing the difference (removing `JPH_DET_LOG` from other places if they cause too much spam).

Specifically for `CharacterVirtual` I'm aware of one thing that will make it non deterministic and that is if `CharacterVirtual::GetMaxHitsExceeded` is true after a step. The problem here is that it discards hits at that point, but the hits are not deterministically ordered because the bodies they come from have been processed on multiple threads. If this is the case you can increase `CharacterVirtualSettings::mMaxNumHits` or ensure the geometry that the character walks on is less complex (256 simultaneous hits for a character is pretty slow already). I can probably also make a code fix for this (at the cost of some more CPU cycles) when `JPH_CROSS_PLATFORM_DETERMINISTIC` is defined.

**Comment by jrouwe** [comment]
B.t.w. If you have a repro case that you can e.g. insert into the `HelloWorld` application then I can take a look at it. I've done this process quite often so I usually find the problem quite quickly.

**Reply by AndreaCatania** [reply]
Thanks a lot for the help. I'm currently trying to enable `JPH_ENABLE_DETERMINISM_LOG` to identify exactly what is breaking. This information should significantly simplify finding the issue.

I don't think `mMaxNumHit` is causing the problem because the kinematic body is only touching the ground, and `mMaxNumHit` is set to the default value of 256.

Reproducing the issue in the Hello World example would be very interesting, but honestly, I find it strange that you haven't encountered a determinism issue before. Therefore, I tend to believe that the problem is somehow caused by my setup or configuration flags, and I believe it would work fine in the Hello World example.

I'll keep trying to reproduce it at some point, but in the meantime, I wonder if you see anything unusual in my CMake configuration or Virtual Character processing.

**Reply by jrouwe** [reply]
CharacterVirtual is currently not being tested as part of the cross platform determinism checks, so I wouldn't be surprised if it has an issue.

**Comment by AndreaCatania** [comment]
Hello @jrouwe,

I have been able to collect some more information and verify that the determinism issue occurs on a single byte, specifically related to the contact "distance."

Following **is the initial state received from the linux server** and before processing `MoveCharacter` on the windows client:
```
- GroundState: 0
- GroundID: 636
- GroundShapeID: 4294967295
- GroundPosition: X: 0xc050d76376431223 Y: 0xbfe122e3ca2fcf30 Z: 0xc035f9a85c7ba3e5
- GroundNorm: X: 0x0 Y: 0x3f800000 Z: 0x0
- GroundVelocity: X: 0x0 Y: 0x0 Z: 0x0
- Pos: X: 0xc050d7637a431223 Y: 0xbfe122e44a2fcf30 Z: 0xc035f9a88c7ba3e5
- Rot: X: 0x0 Y: 0x0 Z: 0x0 W: 0x3f800000
- Vel: X: 0x0 Y: 0x0 Z: 0x0
- LastDeltaTime: 0x3c888889
- mMaxHitsExceeded: false
- ContactsNum: 1
-  Contact
-   |- Pos: X: 0xc050d76376431223 Y: 0xbfe122e3ca2fcf30 Z: 0xc035f9a85c7ba3e5
-   |- Vel: X: 0x0 Y: 0x0 Z: 0x0
-   |- ContactNormal: X: 0x0 Y: 0x3f800000 Z: 0x0
-   |- SurfaceNormal: X: 0x0 Y: 0x3f800000 Z: 0x0
-   |- Distance : 0xb4540000
-   |- Fraction: 0x0
-   |- BodyB: 636
-   |- SubShapeB: 4294967295
-   |- MotionTypeB: 0
-   |- HadCollision: true
-   |- WasDiscarded: false
-   |- CanPushChar: true
--------------------
 Saved state: 0, 0, 0, 0, 124, 2, 0, 0, 255, 255, 255, 255, 35, 18, 67, 118, 99, 215, 80, 192, 48, 207, 47, 202, 227, 34, 225, 191, 229, 163, 123, 92, 168, 249, 53, 192, 0, 0, 0, 0, 0, 0, 128, 63, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 35, 18, 67, 122, 99, 215, 80, 192, 48, 207, 47, 74, 228, 34, 225, 191, 229, 163, 123, 140, 168, 249, 53, 192, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 63, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 137, 136, 136, 60, 0, 1, 0, 0, 0, 35, 18, 67, 118, 99, 215, 80, 192, 48, 207, 47, 202, 227, 34, 225, 191, 229, 163, 123, 92, 168, 249, 53, 192, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 63, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 63, 0, 0, 0, 0, 0, 0, 84, 180, 0, 0, 0, 0, 124, 2, 0, 0, 255, 255, 255, 255, 0, 1, 0, 1, 
```

Here the State after processing `MoveCharacter` on the Windows client (`MoveCharacter` implementation is on the first post)
```
- GroundState: 0
- GroundID: 636
- GroundShapeID: 4294967295
- GroundPosition: X: 0xc050d76376431223 Y: 0xbfe122e3ca2fcf30 Z: 0xc035f9a85c7ba3e5
- GroundNorm: X: 0x0 Y: 0x3f800000 Z: 0x0
- GroundVelocity: X: 0x0 Y: 0x0 Z: 0x0
- Pos: X: 0xc050d7637a431223 Y: 0xbfe122e44a2fcf30 Z: 0xc035f9a88c7ba3e5
- Rot: X: 0x0 Y: 0x0 Z: 0x0 W: 0x3f800000
- Vel: X: 0x0 Y: 0x0 Z: 0x0
- LastDeltaTime: 0x3c888889
- mMaxHitsExceeded: false
- ContactsNum: 1
-  Contact
-   |- Pos: X: 0xc050d76376431223 Y: 0xbfe122e3ca2fcf30 Z: 0xc035f9a85c7ba3e5
-   |- Vel: X: 0x0 Y: 0x0 Z: 0x0
-   |- ContactNormal: X: 0x0 Y: 0x3f800000 Z: 0x0
-   |- SurfaceNormal: X: 0x0 Y: 0x3f800000 Z: 0x0
-   |- Distance : 0xb4640000
-   |- Fraction: 0x0
-   |- BodyB: 636
-   |- SubShapeB: 4294967295
-   |- MotionTypeB: 0
-   |- HadCollision: true
-   |- WasDiscarded: false
-   |- CanPushChar: true
--------------------
 Saved state: 0, 0, 0, 0, 124, 2, 0, 0, 255, 255, 255, 255, 35, 18, 67, 118, 99, 215, 80, 192, 48, 207, 47, 202, 227, 34, 225, 191, 229, 163, 123, 92, 168, 249, 53, 192, 0, 0, 0, 0, 0, 0, 128, 63, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 35, 18, 67, 122, 99, 215, 80, 192, 48, 207, 47, 74, 228, 34, 225, 191, 229, 163, 123, 140, 168, 249, 53, 192, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 63, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 137, 136, 136, 60, 0, 1, 0, 0, 0, 35, 18, 67, 118, 99, 215, 80, 192, 48, 207, 47, 202, 227, 34, 225, 191, 229, 163, 123, 92, 168, 249, 53, 192, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 63, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 63, 0, 0, 0, 0, 0, 0, 100, 180, 0, 0, 0, 0, 124, 2, 0, 0, 255, 255, 255, 255, 0, 1, 0, 1, 
 ```

As you can see, the VirtualCharacter state generated by the Linux server has the contact `|- Distance : 0xb4540000`, while on the Windows client it outputs `|- Distance : 0xb4640000` even after resetting the state. If you compare the binary buffer obtained with `SavedState` (last line of each report), you can notice that only a single byte is different: Linux outputs `84`, while Windows outputs `100`.

I'll keep investigating and reporting any updates. If anything comes to mind or you need anything from my side, let me know.

**Reply by jrouwe** [reply]
I would start annotating the inputs to `CharacterVirtual::CheckCollision` and the contacts that are being provided in `CharacterVirtual::ContactCastCollector::AddHit`.

Given that `Distance : 0xb4640000` is a very small floating point number, the only other thing that I would double check is if the floating point state (denormals are zero, flush to zero, floating point rounding mode) are the same on client and server on all threads that execute Jolt code.

B.t.w. does this scene only contain the static world and a CharacterVirtual? Or are there other dynamic objects in there that could affect the character (it would be good to rule those out by ensuring that the world is completely static)? And what kind of shapes is the character moving over?

**Reply by AndreaCatania** [reply]
Right now, the world is composed of three bodies:
- **1 static body**, which is the ground (using a scaled box `[1,1,1]`)
- **1 kinematic body** (the character using a capsule)
- **1 rigid body** using a sphere

Something **very important** that I've noticed is that the byte difference on the `Distance` variable also occurs between the shipped build and the in-editor build on Windows. So I don't need to log on linux to trigger it.
The editor build in Windows is outputting `84` (like on the Linux server), while the shipped build outputs `100`, still on Windows.

Initially, I thought it had something to do with the shape size, as the Unreal cooker might introduce some differences in scaling, etc., but that's not the case because the shape sizes and scales are consistent across builds.

Another thing to note is that there is no difference between the shipped build and the editor build for Jolt, as the library is compiled with the exact same properties, moreover the jolt lib is not even recompiled during the cooking phase so it's actually the same code.

**Reply by AndreaCatania** [reply]
I've finally find the trigger of the issue and it's indeed the unreal engine level cooker! 

Ground position on Linux server and in editor builds
> Pos: X: 0xc01ed261e0fd2dcd Y: 0xbff09172070ee60a Z: 0x403a0d6327f81d`94`
> Pos: -7.70545, -1.03551, 26.0523

Ground position on shipped build Windows
> Pos: X: 0xc01ed261e0fd2dcd Y: 0xbff09172070ee60a Z: 0x403a0d6327f81d`95`
> Pos: -7.70545, -1.03551, 26.0523

It introduces a discrepancy in the object position between the in-editor, Linux server, Windows client (and probably even the Linux client).

As you can see above, the discrepancy is minimal and is only visible at the byte level. It actually makes sense that the entire simulation behaved non-deterministically and isn't related solely to the Virtual Character.

Now, the question is how to solve it. Clearly, I can't make the Unreal Engine cooker deterministic, and I can't network the state for **all** the static bodies either. I can create a custom cooker for the bodies, but that is challenging. I wonder whether rounding all floats used to create the bodies and shapes to the nearest `0.001` is a great long-term solution.

Do you think that would be a good solution? If not, do you have any better ideas?

**Reply by jrouwe** [reply]
I don't really know how the Unreal cooker works and why it would result in a difference in the lowest bit. Shouldn't it write out the cooked data and then use the same data on all platforms, including the editor build? The actual difference of 1 bit sounds like the kind of error you get from using a fused multiply add vs not using one. But that would only go wrong if the cooking code is different between target platforms.

Rounding to the nearest 0.001 has an edge case if the client's position is e.g. 0.000499999..9 and the server is 0.00050000..1, so a single bit difference can still result in a difference. A hacky solution that doesn't suffer from this problem is to set the last byte of all positions to 0x80 before cooking and to set it to 0x00 after cooking so that any rounding up/down will not affect the higher bytes. But to me it sounds like you should first investigate why the cooking produces different results to begin with.

**Reply by AndreaCatania** [reply]
Ok, let me investigate that, thanks!

**Reply by AndreaCatania** [reply]
Thanks a lot for recommending to dig deeper. I've actually found that the problem is not the cooker but the unit conversion from centimeters (Unreal Engine unit) to meters (Jolt).

```cpp
UnitScale::UnrealToJolt(ConvertR(InitialTransform.GetTranslation()));
// Pos: X: 0xc01ed261e0fd2dcd Y: 0xbff09172070ee60a Z: 0x403a0d6327f81d95

inline JPH::RVec3 ConvertR(const FVector& V)
{
	return JPH::RVec3(V.X, V.Z, V.Y);
}

template <class T>
inline T UnrealToJolt(T D)
{
	D /= 100.0;
	return D;
}
```

The function `ConvertR` is used to swap the axes from left-handed to right-handed, and `UnrealToJolt` is used to scale the measurements from centimeters to meters.

The interesting part is that the issue only occurs when everything is written in a single line, as shown above. When I split that line into three separate lines like below, the produced value is consistent with what I get on the in editor build and linux.

```cpp
UE_LOG(LogTemp, Error, TEXT("------------> X: %s, Y: %s, Z: %s"), UTF8_TO_TCHAR(FLASH::float_to_hex_string(v.X).c_str()), UTF8_TO_TCHAR(FLASH::float_to_hex_string(v.Y).c_str()), UTF8_TO_TCHAR(FLASH::float_to_hex_string(v.Z).c_str()));
const JPH::RVec3 v2 = ConvertR(v);
UE_LOG(LogTemp, Error, TEXT("------------> %s"), UTF8_TO_TCHAR(FLASH::to_hex_string(v2).c_str()));
const JPH::RVec3 v3 = UnitScaler::UnrealToJolt(v2);
UE_LOG(LogTemp, Error, TEXT("------------> %s"), UTF8_TO_TCHAR(FLASH::to_hex_string(v3).c_str()));
const JPH::RVec3 Position = UnitScaler::UnrealToJolt(ConvertR(InitialTransform.GetTranslation()));
// Pos: X: 0xc01ed261e0fd2dcd Y: 0xbff09172070ee60a Z: 0x403a0d6327f81d94
```

I think that the optimization level used to compile the game during cooking on Windows might change the way the division is performed, leading to this discrepancy. I've tried wrapping the `ConvertR` and `UnrealToJolt` functions directly on their header with `JPH_PRECISE_MATH_ON` and `JPH_PRECISE_MATH_OFF`, but it doesn't make much difference. I wonder, how I can ensure that the division is consistent without using `JPH_PRECISE_MATH_ON` in every file?

**Reply by AndreaCatania** [reply]
Actually, `JPH_PRECISE_MATH_ON` didn't solve the issue because `|| defined(JPH_CROSS_PLATFORM_DETERMINISTIC)` is disabling it in the Unreal Engine context. Disabling it in Unreal Engine is incorrect, so what I did was extract that section without including that define, and now the `UnitConverter` produces the correct value during the division. I've also moved these functions into a `.cpp` file, avoiding the `inline` keyword, which could cause the pragma to be skipped in some cases.


<details>
<summary>Ported version of JPH_PRECISE_MATH_ON</summary>

```cpp
// Ported from Jolt
// Macro to enable floating point precise mode and to disable fused multiply add instructions
#if defined(JPH_COMPILER_GCC)
	#define FLASH_PRECISE_MATH_ON \
		_Pragma("GCC push_options") \
		_Pragma("GCC optimize(\"-fno-fast-math\")") \
		_Pragma("GCC optimize(\"-fno-fp-contract\")")
	#define FLASH_PRECISE_MATH_OFF \
		_Pragma("GCC pop_options")
#elif defined(JPH_COMPILER_CLANG)
// We compile without -ffast-math because pragma float_control(precise, on) doesn't seem to actually negate all of the -ffast-math effects and causes the unit tests to fail (even if the pragma is added to all files)
// On clang 14 and later we can turn off float contraction through a pragma (before it was buggy), so if FMA is on we can disable it through this macro
#if (defined(JPH_CPU_ARM) && !defined(JPH_PLATFORM_ANDROID) && __clang_major__ >= 16) || (defined(JPH_CPU_X86) && __clang_major__ >= 14)
		#define FLASH_PRECISE_MATH_ON						\
			_Pragma("float_control(precise, on, push)")	\
			_Pragma("clang fp contract(off)")
		#define FLASH_PRECISE_MATH_OFF					\
			_Pragma("float_control(pop)")
#elif __clang_major__ >= 14 && (defined(JPH_USE_FMADD) || defined(FP_FAST_FMA))
		#define FLASH_PRECISE_MATH_ON						\
			_Pragma("clang fp contract(off)")
		#define FLASH_PRECISE_MATH_OFF					\
			_Pragma("clang fp contract(on)")
#else
		#define FLASH_PRECISE_MATH_ON
		#define FLASH_PRECISE_MATH_OFF
#endif
#elif defined(JPH_COMPILER_MSVC)
// Unfortunately there is no way to push the state of fp_contract, so we have to assume it was turned on before FLASH_PRECISE_MATH_ON
#define FLASH_PRECISE_MATH_ON							\
		__pragma(float_control(precise, on, push))		\
		__pragma(fp_contract(off))
#define FLASH_PRECISE_MATH_OFF						\
		__pragma(fp_contract(on))						\
		__pragma(float_control(pop))
#else
	#error Undefined
#endif
```

</details>

I'm now conducting more tests to verify that everything else is working fine. Please let me know if you think there is a better way to handle this `UnitScaler::UnrealToJolt(ConvertR(InitialTransform.GetTranslation()));`.

**Reply by jrouwe** [reply]
I would expect that if you write your function as:

```
template <class T>
inline T UnrealToJolt(T D)
{
	D *= 0.01;
	return D;
}
```

that you don't need any of this. 

With `/fp:precise` `D /= 100.0` stays as is, without `/fp:precise` it will be rewritten to `D *= 0.01` which is slightly different. With the code above, the resulting assembly should be the same in both cases.

**Reply by AndreaCatania** [reply]
That's nice, thanks!
